backend/src/agents/web_agent.py

```diff
@@ -15,48 +15,53 @@
 engine = PromptEngine()
 
 
-def web_general_search_core(search_query, llm, model) -> str:
+async def web_general_search_core(search_query, llm, model) -> str:
     try:
         search_result = perform_search(search_query, num_results=15)
         if search_result["status"] == "error":
             return "No relevant information found on the internet for the given query."
         urls = search_result["urls"]
         logger.info(f"URLs found: {urls}")
         for url in urls:
-            content = perform_scrape(url)
+            content = await perform_scrape(url)
             if not content:
                 continue
-            summary = perform_summarization(search_query, content, llm, model)
+            summary = await perform_summarization(search_query, content, llm, model)
             if not summary:
                 continue
-            if is_valid_answer(summary, search_query):
+            if await is_valid_answer(summary, search_query):
                 return summary
         return "No relevant information found on the internet for the given query."
     except Exception as e:
         logger.error(f"Error in web_general_search_core: {e}")
         return "An error occurred while processing the search query."
 
+
 @tool(
     name="web_general_search",
-    description="Search the internet based on the query provided and then "
-                "get the meaningful answer from the content found",
+    description=(
+        "Search the internet based on the query provided and then get the meaningful answer from the content found"
+    ),
     parameters={
         "search_query": Parameter(
             type="string",
             description="The search query to find information on the internet",
         ),
     },
 )
-def web_general_search(search_query, llm, model) -> str:
-    return web_general_search_core(search_query, llm, model)
+async def web_general_search(search_query, llm, model) -> str:
+    return await web_general_search_core(search_query, llm, model)
+
 
 def get_validator_agent() -> Agent:
     return ValidatorAgent(config.validator_agent_llm, config.validator_agent_model)
 
-def is_valid_answer(answer, task) -> bool:
-    is_valid = (get_validator_agent().invoke(f"Task: {task}  Answer: {answer}")).lower() == "true"
+
+async def is_valid_answer(answer, task) -> bool:
+    is_valid = (await get_validator_agent().invoke(f"Task: {task}  Answer: {answer}")).lower() == "true"
     return is_valid
 
+
 def perform_search(search_query: str, num_results: int) -> Dict[str, Any]:
```

Comment:
Any reason why this function i.e. perform_search is nor async?

Reply:
In most cases I've only made functions async if they use something that needs to be awaited. (There are some exceptions but those are to do with inheriting a method that is expected to be async e.g the `mock_llm` `chat` method). The search_urls method we call here makes use of the `googlesearch` library which as far as I can see does not have support for async.